src/net/behaviour.rs

```diff
@@ -328,25 +333,31 @@ impl<P: StoreParams> NetworkBehaviourEventProcess<GossipsubEvent> for NetworkBac
     fn inject_event(&mut self, event: GossipsubEvent) {
         match event {
             GossipsubEvent::Message {
-                message: GossipsubMessage { data, topic, .. },
+                message:
+                    GossipsubMessage {
+                        data,
+                        topic,
+                        source,
+                        ..
+                    },
+                propagation_source,
                 ..
             } => {
-                if let Some(subscribers) = self.subscriptions.get_mut(topic.as_str()) {
-                    subscribers
-                        .retain(|subscriber| subscriber.unbounded_send(data.clone()).is_ok());
-                    if subscribers.is_empty() {
-                        self.unsubscribe(topic.as_str());
-                        self.subscriptions.remove(topic.as_str());
-                    }
-                }
+                self.notify_subscribers(
+                    &*topic.to_string(),
+                    GossipEvent::Message(source.unwrap_or(propagation_source), data),
```

Comment:
under which conditions is the `source` empty?

Reply:
1. If there is not validation done, the source field is ignored, see https://github.com/libp2p/rust-libp2p/blob/c98b9ef407575b4c3d5e06d0a9500850b21df8f3/protocols/gossipsub/src/protocol.rs#L304-L315. However, we're using `ValidationMode::Permissive`, so if the source field is present in the message, it is validated (needs to be a valid `PeerId`) and passed along.
2. Gossipsub is an addition to the libp2p pubsub rpcs, see https://github.com/libp2p/specs/blob/master/pubsub/README.md#the-message. I guess they made it optional to have on protobuf to rule them all.

So, in the use-cases we encounter, `source` should always be `Some`.

Reply:
Thanks for the explanation!